qui/tray/devices.py

```diff
@@ -132,321 +66,267 @@ def attach_item(self, menu_item):
+            vm_to_attach.devices[menu_item.device.devclass].attach(assignment)
+
+            self.gtk_app.emit_notification(
+                "Attaching device",
+                "Attaching {} to {}".format(self.device.description,
+                                            menu_item.vm),
+                Gio.NotificationPriority.NORMAL)
+        except qubesadmin.exc.QubesException as ex:
+            self.gtk_app.emit_notification(
+                "Error",
+                "Attaching device {0} to {1} failed. Error: {2}".format(
+                    self.device.description, menu_item.vm, ex),
+                Gio.NotificationPriority.HIGH,
+                error=True)
         except Exception:  # pylint: disable=broad-except
             traceback.print_exc(file=sys.stderr)
 
     def detach_item(self):
-        for menu_item in self.attached_items:
-            emit_notification(self.gtk_app,
-                              "Detaching device",
-                              "Detaching {} from {}".format(
-                                  self.device.description,
-                                  menu_item.vm.name),
-                              Gio.NotificationPriority.NORMAL)
-            for assignment\
-                    in menu_item.vm.devices[menu_item.devclass].assignments():
-                try:
-                    if assignment.device == self.device:
-                        menu_item.vm.devices[menu_item.devclass].detach(
-                            assignment)
-                except exc.QubesException as ex:
-                    emit_notification(self.gtk_app,
-                                      "Error",
-                                      "Detaching device {0} from {1} failed. "
-                                      "Error: {2}".format(
-                                          self.device.description,
-                                          menu_item.vm, ex),
-                                      Gio.NotificationPriority.HIGH,
-                                      error=True)
+        for vm in self.device.attachments:
+            self.gtk_app.emit_notification(
+                "Detaching device",
+                "Detaching {} from {}".format(self.device.description, vm),
+                Gio.NotificationPriority.NORMAL)
+            try:
+                assignment = qubesadmin.devices.DeviceAssignment(
+                    self.device.backend_domain, self.device.ident,
+                    persistent=False)
+                self.qapp.domains[vm].devices[self.device.devclass].detach(
+                    assignment)
+            except qubesadmin.exc.QubesException as ex:
+                self.gtk_app.emit_notification(
+                    "Error",
+                    "Detaching device {0} from {1} failed. Error: {2}".format(
+                        self.device.description, vm, ex),
+                    Gio.NotificationPriority.HIGH,
+                    error=True)
```

Comment:
This is inconsistent. Since this method is also used in `attach_item()`, it should give some indication whether detach succeeded or not. Otherwise, `attach_item()` will try to attach even if detach failed (leading to device being attached to multiple VMs). The purpose of `detach_item()` call in `attach_item()` was to avoid this situation.

On the other hand, the use case of switching device to another VM is (I think) rare, compared to simply detaching it.  So, maybe the UI should be optimized for this common case of just detaching the device? In this case, I'd list only VM(s) to which device is currently connected, to ease clicking on it.
@andrewdavidwong, is my assumption correct here?

Reply:
> On the other hand, the use case of switching device to another VM is (I think) rare, compared to simply detaching it. So, maybe the UI should be optimized for this common case of just detaching the device? In this case, I'd list only VM(s) to which device is currently connected, to ease clicking on it.
> @andrewdavidwong, is my assumption correct here?

In my experience, it's not extremely common, but I'm not sure I'd call it "rare." I think it's reasonable to optimize for the most common "detaching" use case, but some users will certainly want to switch devices to other VMs. Perhaps the VM to which the device is currently connected should be visually more prominent (e.g., shown at the top of a list, with a separator below it, and all the other VMs below that).

Reply:
I've changed it to be bolded - to be honest I'm a bit wary about changing the VM order, it makes it harder to find the one you're looking for. I can be convinced otherwise, but as for now it seems for me that bolding it made it quite prominent enough.